### PR TITLE
Support per listener job classes

### PR DESCRIPTION
**Pull request by bessey**
Adds support for a listener to opt into being its own async job, rather
than using the generic Wisper job.

This can help with observability and prioritisation, since you can give
different listeners different priorities, and they can be distinguished
in APM software with ActiveJob integrations such as Datadog.

You can see how this API hides much of the complexity from the application in the API PR https://github.com/ImmersiveLabsUkOrg/immersive-labs-api/pull/3597

### CODE REVIEW

File: lib/wisper/active_job/listener.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Wisper
+  module ActiveJob
+    # Helper class to allow an ActiveJob to be its own asynchronous listener without Wisper's
+    # Wisper::ActiveJobBroadcaster::Wrapper being needed
+    module Listener
+      def self.included(base)
+        return if base < ::ActiveJob::Base
+        raise("Wisper::ActiveJob::Listener has been included in a class which does not extend from ActiveJob::Base. " \
+          "Please ensure your class has ActiveJob::Base in its ancestry (either directly or via e.g. ApplicationJob)")
```
Comment (hvenables) [inline lib/wisper/active_job/listener.rb:9]: So this enforces that every subscriber that wants to perform async has to inherit from active job

File: lib/wisper/active_job_broadcaster.rb [hunk 2]
```diff
@@ -5,7 +5,11 @@
 module Wisper
   class ActiveJobBroadcaster
     def broadcast(subscriber, publisher, event, args)
-      Wrapper.perform_later(subscriber.name, event, args)
+      if subscriber < ActiveJob::Listener
+        subscriber.perform_later(event, args)
+      else
+        Wrapper.perform_later(subscriber.name, event, args)
+      end
```
Comment (hvenables) [inline lib/wisper/active_job_broadcaster.rb:12]: If we have enforced everyone to use there own class do we need this? and do we need the wrapper?

**Review by hvenables — COMMENTED** [review]
Looking good just a couple of comments to clarify my understanding but ready to approve once i understand better :)

File: lib/wisper/active_job_broadcaster.rb (hunk 2 above)
Comment (bessey) [reply lib/wisper/active_job_broadcaster.rb:12]: We have not enforced that. In the current implementation, `Wisper::ActiveJob::Listener` is optional, so we still need the Wrapper for classes that don't include `Wisper::ActiveJob::Listener`.

For one, if this is going to get upstreamed, we have to maintain backwards compatibility, and even in our own internal PR we aren't using this everywhere from the get-go, so we also need that backwards compat.

Make sense?

Comment (hvenables) [reply lib/wisper/active_job_broadcaster.rb:12]: :+1:

**Review by hvenables — APPROVED** [review]
LGTM

File: README.md [hunk 3]
```diff
@@ -28,17 +28,34 @@ Additionally, you should also ensure that your methods used to handle events und
...
 ```ruby
 class MyListener
-  def self.event_name
+  def self.event_name(arg1, arg2)
   end
 end
 ```
 
+If you prefer to have a job class for each event listener, have your listener extend from `ActiveJob::Base`, include
+`Wisper::ActiveJob::Listener` it, and Wisper will automatically use this job instead of its own internal
```
Comment (sam-kayg) [inline README.md]:
```suggestion
`Wisper::ActiveJob::Listener` in it, and Wisper will automatically use this job instead of its own internal
```

**Review by sam-kayg — APPROVED** [review]

**Pull request merged**